soufi/finders/yum.py

```diff
@@ -224,3 +220,69 @@ def populate_archive(self, *args, **kwargs):  # pragma: no cover
 
     def __repr__(self):
         return self.urls[0]
+
+
+# NOTE(nic): repomd objects require extra-special care and handling.  They
+#  are thin wrappers around ElementTree objects, which means they use an
+#  obnoxious amount of memory, are notoriously hostile to being pickled,
+#  and are also averse to being automagically garbage-collected when they go
+#  out of scope.  This puts them at odds with our caching strategy,
+#  and makes them less-than-ideal for use inside long-running processes.
+#
+#  To deal with the dearth of effective memory management, we will run
+#  all repomd lookups in subprocesses, that will return any/all "needles"
+#  found in the "haystacks" we provide.  This will let the OS
+#  efficiently reclaim all the pages used upon completion.
+def do_task(target, *args):
+    """Run the target callable in a subprocess and return its response."""
+    queue = Queue()
+    process = Process(target=target, args=(queue,) + args)
+    process.start()
+    response = queue.get()
+    if process.is_alive():
+        process.terminate()
+    return response
+
+
+# NOTE(nic): To allow for serializing repomd objects, We will instead
+#  re-serialize the object into its original XML, then re-create and use a
+#  new Repo object from the cached XML on every cache hit.  This makes cache
+#  hits relatively expensive, but they're still orders of magnitude faster
+#  than the alternative.  The XML is LZMA-compressed here to keep cache
+#  storage utilization low, and in the case of Redis caching, to reduce the
+#  amount of cache traffic sent over the wire.
+def get_repomd(queue, url):
+    try:
+        repo = repomd.load(url)
+    except urllib.error.HTTPError:
+        queue.put((None, None))
+        return
+    payload = repomd.defusedxml.lxml.tostring(repo._metadata)
+    queue.put((str(repo.baseurl), lzma.compress(payload)))
```

Comment:
This is missing the URL cleanup from the previous iteration, and the functests fall over because of it.  Will fix.